contrib/src/main/java/com/datatorrent/contrib/cassandra/CassandraPOJOInputOperator.java

```diff
@@ -359,24 +373,43 @@ public void emitTuples()
   {
     super.emitTuples();
     if (lastRowInBatch != null) {
-      switch (primaryKeyColumnType.getName()) {
-        case INT:
-          startRow = lastRowInBatch.getInt(primaryKeyColumn);
-          break;
-        case COUNTER:
-          startRow = lastRowInBatch.getLong(primaryKeyColumn);
-          break;
-        case FLOAT:
-          startRow = lastRowInBatch.getFloat(primaryKeyColumn);
-          break;
-        case DOUBLE:
-          startRow = lastRowInBatch.getDouble(primaryKeyColumn);
-          break;
-        default:
-          throw new RuntimeException("unsupported data type " + primaryKeyColumnType.getName());
-      }
+      startRowToken = getPrimaryKeyToken(primaryKeyColumnType.getName());
     }
+  }
 
+  private Long getPrimaryKeyToken(DataType.Name primaryKeyDataType)
+  {
+    Object keyValue;
+    switch (primaryKeyDataType) {
+    case UUID:
+      keyValue = lastRowInBatch.getUUID(primaryKeyColumn);
+      break;
+    case INT:
+      keyValue = lastRowInBatch.getInt(primaryKeyColumn);
+      break;
+    case COUNTER:
+      keyValue = lastRowInBatch.getLong(primaryKeyColumn);
+      break;
+    case FLOAT:
+      keyValue = lastRowInBatch.getFloat(primaryKeyColumn);
+      break;
+    case DOUBLE:
+      keyValue = lastRowInBatch.getDouble(primaryKeyColumn);
+      break;
+    default:
+      throw new RuntimeException("unsupported data type " + primaryKeyColumnType.getName());
+    }
+    return fetchKeyTokenFromDB(keyValue);
+  }
+
+  private Long fetchKeyTokenFromDB(Object keyValue)
+  {
+    PreparedStatement statement = store.getSession().prepare(TOKEN_QUERY);
```

Comment:
Can you move the relevant changes from #186 to this one? In case this is ever cherry-picked, it may leave these changes as is, which might not be very efficient.